sdk/src/slot_history.rs

```diff
@@ -31,11 +46,37 @@ pub enum Check {
 
 impl SlotHistory {
     pub fn add(&mut self, slot: Slot) {
-        for skipped in self.next_slot..slot {
-            self.bits.set(skipped % MAX_ENTRIES, false);
+        if slot > self.next_slot && slot - self.next_slot >= MAX_ENTRIES {
+            // Wrapped past current history,
+            // clear entire bitvec.
+            let full_blocks = (MAX_ENTRIES as usize) / 64;
+            for i in 0..full_blocks {
+                self.bits.set_block(i, 0);
+            }
+            self.next_slot = slot + 1;
+        } else if slot >= self.next_slot {
+            // Current and next pointers are within
+            // the same MAX_ENTRIES slots
+            let start = self.next_slot % MAX_ENTRIES;
+            let end = slot % MAX_ENTRIES;
+            if end >= start {
+                // Straightforward clear between old head and new head
+                for skipped in start..end {
+                    self.bits.set(skipped, false);
+                }
+            } else {
+                // Wrapped case, need to clear the old head to the end
+                // and from 0 to the new head.
+                for skipped in 0..end {
+                    self.bits.set(skipped, false);
+                }
+                for skipped in start..MAX_ENTRIES {
+                    self.bits.set(skipped, false);
+                }
+            }
+            self.next_slot = slot + 1;
         }
         self.bits.set(slot % MAX_ENTRIES, true);
-        self.next_slot = slot + 1;
```

Comment:
nits: This line can be restored because the newly added `if` clauses both have the same code at end.